src/Views/DirectPhotoPage.vala

```diff
@@ -171,6 +192,97 @@ public class DirectPhotoPage : EditingHostPage {
         return true;
     }
 
+    private void populate_external_app_menu (Gtk.Menu menu, PhotoFileFormat file_format, bool raw) {
+        SortedList<AppInfo> external_apps;
+        string[] mime_types;
+
+        foreach (Gtk.Widget item in menu.get_children ()) {
+            menu.remove (item);
+        }
+
+        // get list of all applications for the given mime types
+        mime_types = file_format.get_mime_types ();
+
+        if (!raw) {
+            var files_appinfo = AppInfo.get_default_for_type ("inode/directory", true);
+
+            var files_item_icon = new Gtk.Image.from_gicon (files_appinfo.get_icon (), Gtk.IconSize.MENU);
+            files_item_icon.pixel_size = 16;
+
+            var menuitem_grid = new Gtk.Grid ();
+            menuitem_grid.add (files_item_icon);
+            menuitem_grid.add (new Gtk.Label (files_appinfo.get_name ()));
+
+            var jump_menu_item = new Gtk.MenuItem ();
+            jump_menu_item.add (menuitem_grid);
+
+            var jump_menu_action = AppWindow.get_instance ().lookup_action (AppWindow.ACTION_JUMP_TO_FILE);
+            jump_menu_action.bind_property ("enabled", jump_menu_item, "sensitive", BindingFlags.SYNC_CREATE);
+            jump_menu_item.activate.connect (() => jump_menu_action.activate (null));
+
+            menu.add (jump_menu_item);
+        }
+
+        assert (mime_types.length != 0);
+        external_apps = DesktopIntegration.get_apps_for_mime_types (mime_types);
+
+        foreach (AppInfo app in external_apps) {
+            var menu_item_icon = new Gtk.Image.from_gicon (app.get_icon (), Gtk.IconSize.MENU);
+            menu_item_icon.pixel_size = 16;
+
+            var menuitem_grid = new Gtk.Grid ();
+            menuitem_grid.add (menu_item_icon);
+            menuitem_grid.add (new Gtk.Label (app.get_name ()));
+
+            var item_app = new Gtk.MenuItem ();
+            item_app.add (menuitem_grid);
+
+            item_app.activate.connect (() => {
+                if (raw) {
+                    on_open_with_raw (app.get_commandline ());
+                } else {
+                    on_open_with (app.get_commandline ());
+                }
+            });
+            menu.add (item_app);
+        }
+        menu.show_all ();
+    }
+
+    private void on_open_with (string app) {
+        if (!has_photo ()) {
+            return;
+        }
+
+        try {
+            AppWindow.get_instance ().set_busy_cursor ();
+            get_photo ().open_with_external_editor (app);
+            AppWindow.get_instance ().set_normal_cursor ();
+        } catch (Error err) {
+            AppWindow.get_instance ().set_normal_cursor ();
+            open_external_editor_error_dialog (err, get_photo ());
+        }
+    }
+
+    private void on_open_with_raw (string app) {
```

Comment:
Can you change it so that you get the command_line at the very end, in the future, it'd be even better if we could use that object to spawn the command with arguments
```suggestion
    private void on_open_with_raw (AppInfo app) {
```